File: PresentationTest/PopAdAnimatedTransitioning.m
//
//  PopAdAnimatedTransitioning.m
//  PresentationTest
//
//  Created by Minxin Guo on 5/9/16.
//  Copyright © 2016 aisle411. All rights reserved.
//

#import "PopAdAnimatedTransitioning.h"

@implementation PopAdAnimatedTransitioning
-(NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
    return 0.4;
}

-(void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
    UIViewController *fromVC = [transitionContext viewControllerForKey: UITransitionContextFromViewKey];
    UIViewController *toVC = [transitionContext viewControllerForKey: UITransitionContextToViewControllerKey];
    UIView *fromView = fromVC.view;
    UIView *toView = toVC.view;
    UIView *containerView = [transitionContext containerView];
    
    if (_isPresentation) {
        [containerView addSubview:toView];
    }
    
    UIViewController *animatingVC = _isPresentation ? toVC : fromVC;
    UIView *animatingView = animatingVC.view;
    CGRect finalFrameForVC = [transitionContext finalFrameForViewController:animatingVC];
    CGRect initialFrameForVC = finalFrameForVC;
    initialFrameForVC.origin.y -= initialFrameForVC.size.height;
    
    CGRect initialFrame = _isPresentation ? initialFrameForVC : finalFrameForVC;
    CGRect finalFrame = _isPresentation ? finalFrameForVC : initialFrameForVC;
    animatingView.frame = initialFrame;
    
    
    [UIView animateWithDuration:[self transitionDuration:transitionContext] delay:0 options:0.6 animations:^{
        animatingView.frame = finalFrame;
    } completion:^(BOOL finished) {
        if (!_isPresentation) {
            [fromView removeFromSuperview];
        }
        [transitionContext completeTransition:YES];
    }];
}
@end

File: PresentationTest/PopAdAnimatedTransitioning.swift
//
//  PopAdAnimatedTransitioning.swift
//  PresentationTest
//
//  Created by Minxin Guo on 5/5/16.
//  Copyright © 2016 aisle411. All rights reserved.
//

import UIKit

class PopAdAnimatedTransitioning: NSObject, UIViewControllerAnimatedTransitioning {
    var isPresentation: Bool = false
    
    func transitionDuration(transitionContext: UIViewControllerContextTransitioning?) -> NSTimeInterval {
        return 0.4
    }
    
    func animateTransition(transitionContext: UIViewControllerContextTransitioning) {
        let fromVC = transitionContext.viewControllerForKey(UITransitionContextFromViewControllerKey)
        let toVC = transitionContext.viewControllerForKey(UITransitionContextToViewControllerKey)
        let fromView = fromVC?.view
        let toView = toVC?.view
        let containerView = transitionContext.containerView()
        
        if (isPresentation) {
            containerView?.addSubview(toView!)
        }
        
        let animatingVC = isPresentation ? toVC : fromVC
        let animatingView = animatingVC?.view
        
        let finalFrameForVC = transitionContext.finalFrameForViewController(animatingVC!)
        var initalFrameForVC = finalFrameForVC
        initalFrameForVC.origin.y -= initalFrameForVC.size.height
        
        let initialFrame = isPresentation ? initalFrameForVC : finalFrameForVC
        let finalFrame = isPresentation ? finalFrameForVC : initalFrameForVC
        
        animatingView?.frame = initialFrame
        
        UIView.animateWithDuration(transitionDuration(transitionContext), delay:0, usingSpringWithDamping:0.6, initialSpringVelocity:5.0, options:UIViewAnimationOptions.AllowUserInteraction, animations:{
            animatingView?.frame = finalFrame
            }, completion:{ (value: Bool) in
                if !self.isPresentation {
                    fromView?.removeFromSuperview()
                }
                transitionContext.completeTransition(true)
        })
    }
}
